agent/lib_php_amqplib.c

```diff
@@ -0,0 +1,401 @@
+    /* PHP 8+ */
+    /* gnu compiler needed closure. */
+    ;
+#else
+  /*
+   * Because we had to strdup values to persist them beyond NR_PHP_WRAPPER_CALL,
+   * now we destroy them. There isn't a separate function to destroy all since
+   * some of the params are string literals and we don't want to strdup
+   * everything if we don't have to. RabbitMQ basic_publish PHP 7.x will only
+   * strdup server_address, destination_name, and
+   * messaging_destination_routing_key.
+   */
+  nr_free(message_params.server_address);
+  nr_free(message_params.destination_name);
+  nr_free(message_params.messaging_destination_routing_key);
+#endif
+}
+NR_PHP_WRAPPER_END
+
+/*
+ * PhpAmqpLib\Channel\AMQPChannel::basic_get
+ * Direct access to a queue if no message was available in the queue, return
+ * null
+ *
+ * @param string $queue
+ * @param bool $no_ack
+ * @param int|null $ticket
+ * @throws \PhpAmqpLib\Exception\AMQPTimeoutException if the specified
+ * operation timeout was exceeded
+ * @return AMQPMessage|null
+ */
+NR_PHP_WRAPPER(nr_rabbitmq_basic_get) {
+  zval* amqp_queue = NULL;
+  zval* amqp_exchange = NULL;
+  zval* amqp_routing_key = NULL;
+  zval* amqp_connection = NULL;
+  nr_segment_t* message_segment = NULL;
+  zval** retval_ptr = NR_GET_RETURN_VALUE_PTR;
+
+  nr_segment_message_params_t message_params = {
+      .library = RABBITMQ_LIBRARY_NAME,
+      .destination_type = NR_MESSAGE_DESTINATION_TYPE_EXCHANGE,
+      .message_action = NR_SPANKIND_CONSUMER,
+      .messaging_system = RABBITMQ_MESSAGING_SYSTEM,
+  };
+
+  (void)wraprec;
+
+  amqp_queue = nr_php_get_user_func_arg(1, NR_EXECUTE_ORIG_ARGS);
+  if (nr_php_is_zval_non_empty_string(amqp_queue)) {
+    /* For consumer, this is queue name. */
+    message_params.destination_name
+        = ENSURE_PERSISTENCE(Z_STRVAL_P(amqp_queue));
+  }
+
+  amqp_connection = nr_php_get_zval_object_property(
+      nr_php_execute_scope(execute_data), "connection");
+  /*
+   * In PHP 7.x, the following will create a strdup in
+   * message_params.server_address that needs to be freed.
+   */
+  nr_php_amqplib_get_host_and_port(amqp_connection, &message_params);
+
+  /* Compatibility with PHP 7.x */
+  NR_PHP_WRAPPER_CALL;
+
+  if (NULL == auto_segment) {
+    /*
+     * Must be checked after PHP_WRAPPER_CALL to ensure txn didn't end during
+     * the call.
+     */
+    goto end;
+  }
+  /*
+   *The retval should be an AMQPMessage. nr_php_is_zval_* ops do NULL checks
+   * as well.
+   */
+  if (nr_php_is_zval_valid_object(*retval_ptr)) {
+    /*
+     *  Get the exchange and routing key from the AMQPMessage
+     */
+    amqp_exchange = nr_php_get_zval_object_property(*retval_ptr, "exchange");
+    if (nr_php_is_zval_non_empty_string(amqp_exchange)) {
+      /* Used with consumer only; his is exchange name.  Exchange name is
+       * Default in case of empty string. */
+      message_params.messaging_destination_publish_name
+          = Z_STRVAL_P(amqp_exchange);
+    } else {
```

Comment:
What if `ampq_exchange` is not a string? (This applies to this pattern in any wrappers in this PR).

Reply:
The function param list and/or class definition dictates strings or not.  
nr_php_is_zval_non_empty_string verifies that the value is not null, is not empty, and is a zval string.  Only after that do we do anything with it.

Reply:
Under the hood this is using `zend_read_property()` which can accept many data types.

Reply:
The check nr_php_is_zval_non_empty_string verifies the property is a string before we do anything with it.  If it's not a string, we ignore it.

Reply:
Only set default if it is a valid but empty string
[ead28c2](https://github.com/newrelic/newrelic-php-agent/pull/1009/commits/ead28c21a438e64e0da8dc8dc53a1926246e73df)